lib/IRGen/GenFunc.cpp

```diff
@@ -2425,3 +2425,111 @@ IRGenFunction::createAsyncDispatchFn(const FunctionPointer &fnPtr,
   Builder.CreateRetVoid();
   return dispatch;
 }
+
+void IRGenFunction::emitSuspensionPoint(llvm::Value *toExecutor,
```

Comment:
IRGenFunction has a call to getAsyncExecutor(), currently this always returns the function's executor argument. This is not longer correct after this change.

We should change it so that it sets up the 'current executor' on function entry and that every suspension point resets it.

The call to `coro_suspend_async` returns a tuple of three arguments which are the arguments to the partial function. You can use the return tuple to set the new executor.

```
  %new_current_exectuor = extract_value %suspend_call, #executor_arg_index